### PR TITLE
[Feature] Linearise reward transform

### PR DESCRIPTION
**Pull request by louisfaury**
## Description
Adds a transform to linearise multi-objective rewards into a single one, via a weighted sum. Essentially, this is a reproduction of mo-gym's `LinearReward ` wrapper.

## Motivation and Context
This feature provides a naive way to address multi-objective environments.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: test/test_transforms.py [hunk 1]
```diff
@@ -92,7 +92,9 @@
     TensorSpec,
     TensorStorage,
     Unbounded,
+    UnboundedContinuous,
 )
+from torchrl.data.tensor_specs import BoundedContinuous
```
Comment (louisfaury) [inline test/test_transforms.py:97]: `BoundedContinuous` is not imported at the root of the `data` package -- happy to add to this PR if that's ok.

Comment (vmoens) [reply test/test_transforms.py:97]: Yes that'd be cool!

File: torchrl/envs/transforms/transforms.py [hunk 2]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
         )
         return super().transform_observation_spec(observation_spec)
+
+
+class LineariseRewards(Transform):
+    """Transforms a multi-objective reward signal to a single-objective one via a weighted sum.
+
+    Args:
+        in_keys: The keys under which the multi-objective rewards are found
+        out_keys: The keys under which single-objective rewards should be written. Defaults to `in_keys`.
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9299]:
```suggestion
        out_keys (List[NestedKey], optional): The keys under which single-objective rewards should be written. Defaults to :attr:`in_keys`.
```

File: torchrl/envs/transforms/transforms.py [hunk 3]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
             high=torch.iinfo(torch.int64).max,
         )
         return super().transform_observation_spec(observation_spec)
+
+
+class LineariseRewards(Transform):
+    """Transforms a multi-objective reward signal to a single-objective one via a weighted sum.
+
+    Args:
+        in_keys: The keys under which the multi-objective rewards are found
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9298]:
```suggestion
        in_keys (List[NestedKey]): The keys under which the multi-objective rewards are found.
```

File: torchrl/envs/transforms/transforms.py [hunk 4]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+        out_keys: The keys under which single-objective rewards should be written. Defaults to `in_keys`.
+        weights: Dictates how to weight each reward when summing them. Defaults to `[1.0, 1.0, ...]`.
+
+    Warning:
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9302]:
```suggestion
    .. warning::
```

File: torchrl/envs/transforms/transforms.py [hunk 5]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
         return super().transform_observation_spec(observation_spec)
+
+
+class LineariseRewards(Transform):
+    """Transforms a multi-objective reward signal to a single-objective one via a weighted sum.
+
+    Args:
+        in_keys: The keys under which the multi-objective rewards are found
+        out_keys: The keys under which single-objective rewards should be written. Defaults to `in_keys`.
+        weights: Dictates how to weight each reward when summing them. Defaults to `[1.0, 1.0, ...]`.
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9300]: What's the typing? Is a tensor appropriate? What about a differentiable tensor?

File: torchrl/envs/transforms/transforms.py [hunk 6]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+
+    Warning:
+        If a sequence of `in_keys` of length strictly greater than one is passed (e.g. one group for each agent in a
+        multi-agent set-up), the same weights will be applied for each entry. If you need to aggregate rewards
+        differently for each group, use several `AggregateRewardsTransform` in a row.
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9305]:
```suggestion
        differently for each group, use several :class:`~torchrl.envs.AggregateRewardsTransform` in a row.
```

File: torchrl/envs/transforms/transforms.py [hunk 7]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+    @_apply_to_composite
+    def transform_reward_spec(self, reward_spec: TensorSpec) -> TensorSpec:
+        if not reward_spec.domain == "continuous":
+            raise NotImplementedError(
+                "Aggregation of rewards that take discrete values is not supported."
+            )
+
+        *batch_size, num_rewards = reward_spec.shape
+        weights = (
+            torch.ones(num_rewards, device=reward_spec.device)
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9367]: maybe get the dtype of the reward spec?

File: torchrl/envs/transforms/transforms.py [hunk 8]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+
+    def _apply_transform(self, reward: Tensor) -> TensorDictBase:
+        if self.weights is None:
+            return reward.sum(dim=-1)
+
+        *batch_size, num_rewards = reward.shape
+        num_weights = torch.numel(self.weights)
+        if num_weights != num_rewards:
+            raise ValueError(
+                "The number of rewards and weights should match."
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9402]:
```suggestion
                "The number of rewards and weights should match. "
```

File: torchrl/envs/transforms/transforms.py [hunk 9]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+        weights = (
+            torch.ones(num_rewards, device=reward_spec.device)
+            if self.weights is None
+            else self.weights
+        )
+
+        num_weights = torch.numel(weights)
+        if num_weights != num_rewards:
+            raise ValueError(
+                "The number of rewards and weights should match."
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9375]:
```suggestion
                "The number of rewards and weights should match. "
```

**Review by vmoens — COMMENTED** [review]
LGTM thanks for this!
Just a few suggestions

File: torchrl/envs/transforms/transforms.py (hunk 5 above)
Comment (louisfaury) [reply torchrl/envs/transforms/transforms.py]: I added the typing, `List[float] | Tensor | None`. I didn't give any thoughts for the tensor to be differentiable -- I believe here it would give unexpected results as `weights` are registered as a non-parameter buffer.

File: torchrl/envs/transforms/transforms.py [hunk 10]
```diff
@@ -9287,3 +9289,118 @@ def transform_observation_spec(self, observation_spec: Composite) -> Composite:
...
+        >>> so_env.reward_spec
+        BoundedContinuous(
+            shape=torch.Size([1]),
+            space=ContinuousBox(
+                low=Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.float32, contiguous=True),
+                high=Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.float32, contiguous=True)),
+            ...)
+        >>> td = so_env.rollout(5)
+        >>> td["next", "reward"].shape
+        [5, 1]
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:9328]:
```suggestion
        torch.Size([5, 1])
```

**Review by vmoens — APPROVED** [review]
LGTM thanks!

File: torchrl/envs/transforms/transforms.py (hunk 5 above)
Comment (louisfaury) [reply torchrl/envs/transforms/transforms.py]: Yes -- my understanding about this is that it would therefore be ignored by any optimizer. 
What I mean by 
> it would give unexpected results

is that if you pass a differentiable tensor `weights` to this module in the hope to optimise it somehow, you won't ever get any gradients for it. 

Maybe I'm misunderstanding your comment?

Comment (vmoens) [reply torchrl/envs/transforms/transforms.py]: No no, my comment was just this:
If your weight is differentiable, is the reward going to be differentiable too?
I think it will so that's great (I can imagine some researcher, somewhere, trying to hack something like that)

**Comment by vmoens** [comment]
From the failing tests it seems a `from __future__ import annotations` is missing - can you check @louisfaury?

File: torchrl/envs/transforms/transforms.py (hunk 5 above)
Comment (louisfaury) [reply torchrl/envs/transforms/transforms.py]: Got it -- the answer is that as it, no. 

I also see the value, but maybe we can keep this to anther PR ?
